internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,214 @@
+
+	installEnv := &adiiov1alpha1.InstallEnv{}
+	if err := r.Get(ctx, req.NamespacedName, installEnv); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.ensureISO(ctx, installEnv)
+}
+
+// ensureISO generates ISO for the cluster if needed and will update the condition Reason and Message accordingly.
+// It returns a result that includes ISODownloadURL.
+func (r *InstallEnvReconciler) ensureISO(ctx context.Context, installEnv *adiiov1alpha1.InstallEnv) (ctrl.Result, error) {
+	var inventoryErr error
+	var imageInfo *models.ImageInfo
+	var updatedCluster *common.Cluster
+	imageState := ImageState{
+		State:     ImageCreationError,
+		StateInfo: adiiov1alpha1.ImageStateFailedToCreate,
+	}
+
+	kubeKey := types.NamespacedName{
+		Name:      installEnv.Spec.ClusterRef.Name,
+		Namespace: installEnv.Spec.ClusterRef.Namespace,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+
+	if err := r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		clusterDeploymentRefErr := newKubeAPIError(
+			errors.Wrapf(
+				err,
+				fmt.Sprintf("failed to find clusterDeployment with name %s in namespace %s",
+					installEnv.Spec.ClusterRef.Name, installEnv.Spec.ClusterRef.Namespace)),
+			k8serrors.IsNotFound(err),
+		)
+		return r.updateStatusAndReturnResult(ctx, installEnv, nil, imageState, clusterDeploymentRefErr)
+	}
+
+	c, err := r.Installer.GetClusterByKubeKey(types.NamespacedName{
+		Name:      clusterDeployment.Name,
+		Namespace: clusterDeployment.Namespace,
+	})
+
+	if gorm.IsRecordNotFoundError(err) {
+		inventoryErr = common.NewApiError(http.StatusNotFound, err)
+
+	} else {
+		isoParams := installer.GenerateClusterISOParams{
+			ClusterID:         *c.ID,
+			ImageCreateParams: &models.ImageCreateParams{},
+		}
+		if len(installEnv.Spec.SSHAuthorizedKeys) > 0 {
+			isoParams.ImageCreateParams.SSHPublicKey = installEnv.Spec.SSHAuthorizedKeys[0]
+		}
+		updatedCluster, inventoryErr = r.Installer.GenerateClusterISOInternal(ctx, isoParams)
+	}
+
+	if inventoryErr == nil {
+		imageInfo = updatedCluster.ImageInfo
+		imageState.State = ImageBeingCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, installEnv, imageInfo, imageState, inventoryErr)
+}
+
+func (r *InstallEnvReconciler) updateStatusAndReturnResult(
+	ctx context.Context,
+	installEnv *adiiov1alpha1.InstallEnv,
+	imageInfo *models.ImageInfo,
+	imageState ImageState,
+	err error) (ctrl.Result, error) {
+
+	var res ctrl.Result
+
+	if imageBeingCreated(err) {
+		imageState.State = ImageCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+		// Clear up the error stateInfo while installEnv is being created
+		err = nil
+		r.Log.Infof("Image %s being prepared for cluster %s stateInfo: %s",
+			installEnv.Name, installEnv.ClusterName, imageState.StateInfo)
+	} else if isClientError(err) {
+		imageState.State = ImageCreationError // TODO before merge: should we do this?
```

Comment:
@filanov what do you think of this?
should be set error `State` in this case?

Reply:
I think that specifically for 409 we should mark the image as in progress. 
For other errors lets show them for the user.

Reply:
Ack. So 409 is being handled by [imageBeingCreated()](https://github.com/openshift/assisted-service/pull/1029/files#diff-d6f17279093c7d93d612ed5b1f45cb8ba9893abd5e0fc7bdf26ac7203f2b934dR144) beforehand